tensorflow/compiler/jit/xla_gpu_device.cc

```diff
@@ -52,8 +53,34 @@ Status XlaGpuDeviceFactory::CreateDevices(
     VLOG(1) << "Failed to create XLA_GPU device: " << platform.status();
     return Status::OK();
   }
-
-  for (int i = 0; i < platform.ValueOrDie()->VisibleDeviceCount(); ++i) {
+  const auto& allowed_gpus =
+      session_options.config.gpu_options().visible_device_list();
+  std::unordered_set<int> gpu_ids;
+  int num_visible_devices = platform.ValueOrDie()->VisibleDeviceCount();
+  if (allowed_gpus.empty()) {
+    for (int i = 0; i < num_visible_devices; ++i) gpu_ids.insert(i);
+  } else {
+    const std::vector<string> visible_devices =
+        str_util::Split(allowed_gpus, ',');
+    // copied from gpu/gpu_device.cc Should be redundant since code would fail
+    // there before it gets to here.
+    for (const string& platform_gpu_id_str : visible_devices) {
+      int32 platform_gpu_id;
+      if (!strings::safe_strto32(platform_gpu_id_str, &platform_gpu_id)) {
+        return errors::InvalidArgument(
+            "Could not parse entry in 'visible_device_list': '",
+            platform_gpu_id_str, "'. visible_device_list = ", allowed_gpus);
+      }
+      if (platform_gpu_id < 0 || platform_gpu_id >= num_visible_devices) {
+        return errors::InvalidArgument(
+            "'visible_device_list' listed an invalid GPU id '", platform_gpu_id,
+            "' but visible device count is ", num_visible_devices);
+      }
+      gpu_ids.insert(platform_gpu_id);
+    }
+  }
+  for (int i = 0; i < num_visible_devices; ++i) {
+    if (gpu_ids.count(i) == 0) continue;
```

Comment:
Please just iterate over `gpu_ids`.

Also maybe we should make it an `std::set` so ordering is deterministic.

Reply:
Are you sure that ordering is important here? It is matching the enumeration of GPU ids so simple lookup should be sufficient. I used count for shorthand for find but now I changed it to use find. If you want ordered set, I can use std::set instead of std::unordered_set.

Reply:
> Are you sure that ordering is important here? It is matching the enumeration of GPU ids so simple lookup should be sufficient. 

I was suggesting that we make this a loop over the `gpu_ids` set *and then* make `gpu_ids` an ordered set, so that the behavior of the loop is unchanged.

Reply:
Sorry about that. I misunderstood it. Fixed that now and making sure everything is fine again. Will push the changes shortly.